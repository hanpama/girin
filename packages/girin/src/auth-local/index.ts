import * as argon2 from 'argon2';
import * as jwt from 'jsonwebtoken';

import { Model, field } from '../mongodb';
import { ObjectID } from 'bson';
import AuthLocalModule from './module';
import { globalEnvironment } from '@girin/environment';


function getPasswordSalt() {
  return globalEnvironment.get(AuthLocalModule).PASSWORD_SALT;
}
function getJWTSecretKey() {
  return globalEnvironment.get(AuthLocalModule).JWT_SECRET_KEY;
}

export type BaseUserClass<T extends BaseUser> = typeof BaseUser & {
  new(source: any): T;
}

export class BaseUser extends Model {

  @field()
  protected hashedPassword: string;

  public static async createUserWithPassword(source: { [fieldName: string]: any }, password: string) {

    const hashedPassword = await argon2.hash(password + getPasswordSalt());
    const user = await this.insert({ hashedPassword, ...source });
    return user;
  }

  public authenticate(password: string) {
    return argon2.verify(this.hashedPassword, password + getPasswordSalt());
  }

  public async changePassword(newPassword: string) {
    const hashedPassword = await argon2.hash(newPassword + getPasswordSalt());
    this.hashedPassword = hashedPassword;
    await this.$update({ $set: { hashedPassword } });

    return this;
  }

  public async encodeToken(password: string) {
    const authenticated = await this.authenticate(password);
    if (!authenticated) {
      throw new Error('Authentication Error');
    }
    const { hashedPassword, ...rest } = this.$source;
    return jwt.sign(rest, getJWTSecretKey());
  }

  public static decodeToken<T extends BaseUser>(this: BaseUserClass<T>, token: string): T {
    const result: any = jwt.verify(token, getJWTSecretKey());
    console.log(result);
    const user = new this(result);
    user._id = new ObjectID(result._id);
    return user;
  }
}
