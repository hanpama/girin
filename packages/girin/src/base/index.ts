export * from './scalar';
export * from './Definition';
export * from './MetadataStorage';
export * from './TypeExpression';
export * from './structure';
