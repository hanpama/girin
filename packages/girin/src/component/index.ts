export * from './Component';
export * from './decorators';
