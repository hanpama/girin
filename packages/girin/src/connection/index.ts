export * from './PageInfo';
export * from './Edge';
export * from './Connection';
