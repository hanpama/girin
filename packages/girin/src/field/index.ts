export * from './Field';
export * from './InputField';
// export * from './mount';
