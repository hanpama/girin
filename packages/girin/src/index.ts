export { StringScalar, BooleanScalar, FloatScalar, IntScalar, IDScalar } from './builtins/scalar';
export { List, NonNull } from './type-expression/structure';

export * from './sdl';
export * from './getGraphQLType';
