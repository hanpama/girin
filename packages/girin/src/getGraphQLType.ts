import { MetadataStorage } from "./base/MetadataStorage";
import { DefinitionMetadata } from "./base/DefinitionMetadata";
import { globalMetadataStorage } from "./globalMetadataStorage";


export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(targetClass: Function, options?: TypeBuilderOptions): any {

  const typeMetadata = globalMetadataStorage.getDefinitionMetadata(DefinitionMetadata, targetClass);
  if (!typeMetadata) {
    throw new Error(`Given class ${targetClass.name} has no corresponding metadata`);
  }
  return typeMetadata.typeInstance;
}
