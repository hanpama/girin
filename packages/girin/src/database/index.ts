import { loadPredefinedConnectionFields } from './connection/relay';
import { getGlobalMetadataStorage } from 'girin-typelinker';

loadPredefinedConnectionFields(getGlobalMetadataStorage());

export * from './model';
export * from './connection';
