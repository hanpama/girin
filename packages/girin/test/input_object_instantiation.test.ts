import { graphql, GraphQLSchema, printSchema } from 'graphql';

import { defineType, gql, getGraphQLType } from '../src';


@defineType(gql`
  type Member {
    id: Int!
    username: String!
    email: String!
  }
`)
class Member {
  id: number;
  username: string;
  email: string;
}

@defineType(gql`
  input MemberProfileInput {
    firstName: String!
    lastName: String!
  }
`)
class MemberProfileInput {
  public firstName: string;
  public lastName: string;
}

@defineType(gql`
  input MemberInput {
    username: String!
    email: String!
    profile: MemberProfileInput!
  }
`)
class MemberInput {
  public username: string;
  public email: string;
  public profile: MemberProfileInput;
}

@defineType(gql`
  type Query {
    getMember: Member!
  }
`)
class Query {
  public static getMember(): Member {
    return { id: 1, username: 'Jonghyun', email: 'j@example.com' };
  }
}

@defineType(gql`
  type Mutation {
    instantiateMemberInput(member: MemberInput!): Boolean!
    getEmailFromInput(member: MemberInput!): String!
    getUsernameFromInput(member: MemberInput!): String!
  }
`)
class Mutation {
  public static instantiateMemberInput(source: null, { member }: { member: MemberInput }) {
    const result = member instanceof MemberInput && member.profile instanceof MemberProfileInput;
    return result;
  }
  public static getEmailFromInput(source: null, { member }: {member: MemberInput }) {
    return member.email;
  }

  public static getUsernameFromInput(source: null, { member }: { member: MemberInput }) {
    return member.username;
  }
}

const schema = new GraphQLSchema({
  query: getGraphQLType(Query),
  mutation: getGraphQLType(Mutation),
});


describe('Input object instantiation', () => {
  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('should create InputObjectType instance with given argument value', async () => {
    const result = await graphql({ schema, source: `
      mutation {
        instantiateMemberInput(member: {
          username: "key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
        getUsernameFromInput(member: {
          username: "Key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
        getEmailFromInput(member: {
          username: "Key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
      }
    `});
    expect(result).toEqual({ data: {
      getUsernameFromInput: "Key",
      getEmailFromInput: "k@example.com",
      instantiateMemberInput: true,
    }})
  });
});
