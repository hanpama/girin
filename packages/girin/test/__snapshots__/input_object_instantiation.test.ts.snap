// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input object instantiation generates schema as expected 1`] = `
"type Member {
  id: Int!
  username: String!
  email: String!
}

input MemberInput {
  username: String!
  email: String!
  profile: MemberProfileInput!
}

input MemberProfileInput {
  firstName: String!
  lastName: String!
}

type Mutation {
  instantiateMemberInput(member: MemberInput!): Boolean!
  getEmailFromInput(member: MemberInput!): String!
  getUsernameFromInput(member: MemberInput!): String!
}

type Query {
  getMember: Member!
}
"
`;
